### PR TITLE
Add text-letter-spacing support

### PR DESCRIPTION
**Pull request by pakb**
Add support for https://docs.mapbox.com/mapbox-gl-js/style-spec#layout-symbol-text-letter-spacing

will fix https://github.com/openlayers/ol-mapbox-style/issues/137

### CODE REVIEW

File: stylefunction.js [hunk 1]
```diff
@@ -601,6 +601,19 @@ export default function(olLayer, glStyle, source, resolutions, spriteData, sprit
             label = label.toLowerCase();
           }
           const wrappedLabel = type == 2 ? label : wrapText(label, font, getValue(layer, 'layout', 'text-max-width', zoom, f));
+          const letterSpacing = getValue(layer, 'layout', 'text-letter-spacing', zoom, f);
+          if (letterSpacing && letterSpacing > 0.01 /* default is 0.01 */ ) {
```
Comment (ahocevar) [inline stylefunction.js]: `letterSpacing` should always be available here, so you can remove `letterSpacing && `.

And we should enter this code path only when `letterSpacing >= 0.05`. See below.

File: stylefunction.js [hunk 2]
```diff
@@ -601,6 +601,19 @@ export default function(olLayer, glStyle, source, resolutions, spriteData, sprit
...
+            let wrappedLabelWithLetterSpacing = '';
+            wrappedLabel.split('\n').forEach((line) => {
+              let joinSpaceString = "",
```
Comment (ahocevar) [inline stylefunction.js]: Please rebase your pull request on top of master so the CI catches this, and use single quotes.

File: stylefunction.js [hunk 3]
```diff
@@ -601,6 +601,19 @@ export default function(olLayer, glStyle, source, resolutions, spriteData, sprit
...
+                  spacingCount;
+              for (spacingCount = 0; spacingCount < letterSpacing; spacingCount += 0.1) {
+                joinSpaceString += String.fromCharCode(8202); // Hair space, about 0.1em per character
```
Comment (ahocevar) [inline stylefunction.js]: `'\u200A'` should be faster than `String.fromCharCode(8202)`.

File: stylefunction.js [hunk 4]
```diff
@@ -601,6 +601,19 @@ export default function(olLayer, glStyle, source, resolutions, spriteData, sprit
             label = label.toLowerCase();
           }
           const wrappedLabel = type == 2 ? label : wrapText(label, font, getValue(layer, 'layout', 'text-max-width', zoom, f));
+          const letterSpacing = getValue(layer, 'layout', 'text-letter-spacing', zoom, f);
+          if (letterSpacing && letterSpacing > 0.01 /* default is 0.01 */ ) {
+            let wrappedLabelWithLetterSpacing = '';
+            wrappedLabel.split('\n').forEach((line) => {
```
Comment (ahocevar) [inline stylefunction.js]: For performance reasons, please use a `for` loop instead of `forEach`.

File: stylefunction.js [hunk 5]
```diff
@@ -601,6 +601,19 @@ export default function(olLayer, glStyle, source, resolutions, spriteData, sprit
...
+              let joinSpaceString = "",
+                  spacingCount;
+              for (spacingCount = 0; spacingCount < letterSpacing; spacingCount += 0.1) {
```
Comment (ahocevar) [inline stylefunction.js]: No need to do this in a loop. The `joinSpaceString` can be created with
```js
const joinSpaceString = Array(Math.round(letterSpacing / 0.1)).join('\u200A');
```
This way we add one hair space for a `letterSpacing` of `0.5` to `1.499999`, two for `1.5` to `2.499999` and so on.

Comment (pakb) [reply stylefunction.js]: Isnt't join also part of the filter map reduce, and thus will be a bigger impact as to do a simple for loop?

Comment (ahocevar) [reply stylefunction.js]: You're right regarding performance. And both your and my approach are lame. I've come up with a 3rd way, see https://jsperf.com/pad-strings. So the best is to define a string pool only once that we slice from:
```js
// on the module level
const hairSpacePool = Array(256).join('\u200A');

export function styleFunction(feature, resolution) {
  // ...
  const joinSpaceString = hairSpacePool.slice(0, Math.round(letterSpacing / 0.1));
  // ...
}

**Comment by pakb** [comment]
Should be good, let me know if you think we need to improve something else with this PR

File: stylefunction.js [hunk 6]
```diff
@@ -601,7 +602,18 @@ export default function(olLayer, glStyle, source, resolutions, spriteData, sprit
             label = label.toLowerCase();
           }
           const wrappedLabel = type == 2 ? label : wrapText(label, font, getValue(layer, 'layout', 'text-max-width', zoom, f));
-          text.setText(wrappedLabel);
+          const letterSpacing = getValue(layer, 'layout', 'text-letter-spacing', zoom, f);
+          if (letterSpacing >= 0.05) {
+            let wrappedLabelWithLetterSpacing = '';
+            const wrappedLabelLines = wrappedLabel.split('\n');
+            const joinSpaceString = hairSpacePool.slice(0, Math.round(letterSpacing / 0.1));
+            for (let lineIndex = 0; lineIndex < wrappedLabelLines.length; lineIndex += 1) {
```
Comment (ahocevar) [inline stylefunction.js]:
```suggestion
            for (let l = 0, ll = wrappedLabelLines.length, l < ll; ++l) {
```

File: stylefunction.js [hunk 7]
```diff
@@ -601,7 +602,18 @@ export default function(olLayer, glStyle, source, resolutions, spriteData, sprit
...
+              wrappedLabelWithLetterSpacing += wrappedLabelLines[lineIndex].split('').join(joinSpaceString);
```
Comment (ahocevar) [inline stylefunction.js]:
```suggestion
              wrappedLabelWithLetterSpacing += wrappedLabelLines[l].split('').join(joinSpaceString);
```

Comment (ahocevar) [reply stylefunction.js]: Oh, and one more thing: you need to re-add the line breaks to wrappedLabelWithLetterSpacing.

Comment (pakb) [reply stylefunction.js]: oops, yep forgot that

**Pull request merged**